mp/src/game/client/momentum/mom_system_discord.cpp

```diff
@@ -0,0 +1,512 @@
+    // Check to see if we are joining this map to spectate a player
+    if (m_kJoinSpectateState == WaitOnMap) {
+        m_kJoinSpectateState = WaitOnNone;
+        SpectateTargetFromDiscord();
+    }
+}
+
+// Called each time a map is left
+void CMomentumDiscord::LevelShutdownPreEntity() {
+    ClearDiscordFields(false); // Pass false to retain party/lobby related fields
+    V_strncpy(m_szDiscordState, s_pszInMenusStatusString, DISCORD_MAX_BUFFER_SIZE);
+    V_strncpy(m_szDiscordLargeImageKey, s_pszInMenusLargeImage, DISCORD_MAX_BUFFER_SIZE);
+}
+
+// Called every frame
+void CMomentumDiscord::Update(float frametime) {
+    if (m_iUpdateFrame++ == DISCORD_FRAME_UPDATE_FREQ) {
+        // Stuff that can be done every ~1 second
+        m_iUpdateFrame = 0;
+        UpdateDiscordPartyIdFromSteam();
+        UpdateLobbyNumbers();
+        DiscordUpdate();
+    }
+
+    // Do the rest of this stuff every frame
+    #ifdef DISCORD_DISABLE_IO_THREAD
+        Discord_UpdateConnection();
+    #endif
+    Discord_RunCallbacks();
+}
+
+// Called on game quit
+void CMomentumDiscord::Shutdown() {
+    DevLog("Shutting down Discord\n");
+    Discord_Shutdown();
+}
+
+// -------------------------- Steam API Callbacks --------------------------- //
+
+// When a user joins a lobby
+void CMomentumDiscord::HandleLobbyEnter(LobbyEnter_t* pParam) {
+    if (pParam->m_EChatRoomEnterResponse != k_EChatRoomEnterResponseSuccess || pParam->m_ulSteamIDLobby == 0) {
+        DevWarning("Failed to enter room! Error code: %i\n", pParam->m_EChatRoomEnterResponse);
+        return;
+    }
+
+    m_sSteamLobbyID = CSteamID(pParam->m_ulSteamIDLobby);
+    UpdateDiscordPartyIdFromSteam();
+}
+
+void CMomentumDiscord::HandleLobbyDataUpdate(LobbyDataUpdate_t* pParam) {
+    if (CSteamID(pParam->m_ulSteamIDLobby) != m_sSteamLobbyID) {
+        m_sSteamLobbyID = CSteamID(pParam->m_ulSteamIDLobby);
+        UpdateDiscordPartyIdFromSteam();
+    }
+
+    OnSteamLobbyUpdate();
+}
+
+void CMomentumDiscord::HandleLobbyChatUpdate(LobbyChatUpdate_t* pParam) {
+    if (CSteamID(pParam->m_ulSteamIDLobby) != m_sSteamLobbyID) {
+        m_sSteamLobbyID = CSteamID(pParam->m_ulSteamIDLobby);
+        UpdateDiscordPartyIdFromSteam();
+    }
+
+    OnSteamLobbyUpdate();
+}
+
+// --------------------------- CGameEventListner ---------------------------- //
+
+// Called when there is a game event we are listening for
+// Subscribe to events in `PostInit`
+void CMomentumDiscord::FireGameEvent(IGameEvent* event) {
+    if (!Q_strcmp(event->GetName(), "lobby_leave")) {
+        m_sSteamLobbyID.Clear();
+        m_iDiscordPartySize = 0;
+        m_iDiscordPartyMax = 0;
+        m_szDiscordPartyId[0] = '\0';
+        m_szDiscordJoinSecret[0] = '\0';
+        m_szDiscordSpectateSecret[0] = '\0';
+    }
+}
+
+// ----------------------------- Custom Methods ----------------------------- //
+
+// Returns true if the player is in a map, false otherwise
+bool CMomentumDiscord::InMap() {
```

Comment:
Change this to be a member variable boolean that gets set true on map load, false on level shutdown